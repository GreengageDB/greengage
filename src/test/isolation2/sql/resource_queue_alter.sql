-- Behavioral tests to showcase what happens when resource queue limits are
-- altered concurrently with active and waiting statements.

0:CREATE RESOURCE QUEUE rq_alter WITH (active_statements = 1);
0:CREATE ROLE rq_alter_role RESOURCE QUEUE rq_alter;

--
-- Scenario 1: Bumping the resource queue limit
--
1:SET ROLE rq_alter_role;
1:BEGIN;
1:DECLARE c1 CURSOR FOR SELECT 1;

2:SET ROLE rq_alter_role;
2:BEGIN;
-- The following will block waiting for the resource queue slot.
2&:DECLARE c2 CURSOR FOR SELECT 1;

-- Inflate the limit.
0:ALTER RESOURCE QUEUE rq_alter WITH (active_statements = 2);
-- The limit has been bumped. However we still see that Session 2 is waiting.
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted
  FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders
  FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';

-- New statements actually can get into the queue ahead of Session 2!
1:DECLARE c3 CURSOR FOR SELECT 1;

-- Only once a statement completes (or ERRORs out) is Session 2 woken up, and
-- can complete acquiring the lock for portal c2.
1:CLOSE c1;
2<:
1q:
2q:

--
-- Scenario 2: Decreasing the resource queue limit
--
1:SET ROLE rq_alter_role;
1:BEGIN;
1:DECLARE c1 CURSOR FOR SELECT 1;
1:DECLARE c2 CURSOR FOR SELECT 1;

-- The following will block waiting for the resource queue slot.
2:SET ROLE rq_alter_role;
2:BEGIN;
2&:DECLARE c3 CURSOR FOR SELECT 1;

-- The current situation:
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted
  FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders
  FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';

-- Decrease the limit
0:ALTER RESOURCE QUEUE rq_alter WITH (active_statements = 1);

-- Nothing changes, current holders and waiters continue as they were. However,
-- do note that now we have a situation where rsqcountvalue > rsqcountlimit.
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted
  FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders
  FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';

-- The following will NOT wake up Session 2, as we would exceed the limit if we
-- were to do so.
1:CLOSE c1;
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted
  FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders
  FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';

-- Now this will wake up Session 2
1:CLOSE c2;
2<:

1q:
2q:

0:DROP ROLE rq_alter_role;
0:DROP RESOURCE QUEUE rq_alter;
