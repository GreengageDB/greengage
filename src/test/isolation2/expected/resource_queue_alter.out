-- Behavioral tests to showcase what happens when resource queue limits are
-- altered concurrently with active and waiting statements.

0:CREATE RESOURCE QUEUE rq_alter WITH (active_statements = 1);
CREATE QUEUE
0:CREATE ROLE rq_alter_role RESOURCE QUEUE rq_alter;
CREATE ROLE

--
-- Scenario 1: Bumping the resource queue limit
--
1:SET ROLE rq_alter_role;
SET
1:BEGIN;
BEGIN
1:DECLARE c1 CURSOR FOR SELECT 1;
DECLARE CURSOR

2:SET ROLE rq_alter_role;
SET
2:BEGIN;
BEGIN
-- The following will block waiting for the resource queue slot.
2&:DECLARE c2 CURSOR FOR SELECT 1;  <waiting ...>

-- Inflate the limit.
0:ALTER RESOURCE QUEUE rq_alter WITH (active_statements = 2);
ALTER QUEUE
-- The limit has been bumped. However we still see that Session 2 is waiting.
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
 ?column? | lorrsqname | lorlocktype    | lormode       | lorgranted 
----------+------------+----------------+---------------+------------
 t        | rq_alter   | resource queue | ExclusiveLock | t          
 t        | rq_alter   | resource queue | ExclusiveLock | f          
(2 rows)
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';
 rsqname  | rsqcountlimit | rsqcountvalue | rsqwaiters | rsqholders 
----------+---------------+---------------+------------+------------
 rq_alter | 2             | 1             | 1          | 1          
(1 row)

-- New statements actually can get into the queue ahead of Session 2!
1:DECLARE c3 CURSOR FOR SELECT 1;
DECLARE CURSOR

-- Only once a statement completes (or ERRORs out) is Session 2 woken up, and
-- can complete acquiring the lock for portal c2.
1:CLOSE c1;
CLOSE CURSOR
2<:  <... completed>
DECLARE CURSOR
1q: ... <quitting>
2q: ... <quitting>

--
-- Scenario 2: Decreasing the resource queue limit
--
1:SET ROLE rq_alter_role;
SET
1:BEGIN;
BEGIN
1:DECLARE c1 CURSOR FOR SELECT 1;
DECLARE CURSOR
1:DECLARE c2 CURSOR FOR SELECT 1;
DECLARE CURSOR

-- The following will block waiting for the resource queue slot.
2:SET ROLE rq_alter_role;
SET
2:BEGIN;
BEGIN
2&:DECLARE c3 CURSOR FOR SELECT 1;  <waiting ...>

-- The current situation:
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
 ?column? | lorrsqname | lorlocktype    | lormode       | lorgranted 
----------+------------+----------------+---------------+------------
 t        | rq_alter   | resource queue | ExclusiveLock | t          
 t        | rq_alter   | resource queue | ExclusiveLock | f          
(2 rows)
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';
 rsqname  | rsqcountlimit | rsqcountvalue | rsqwaiters | rsqholders 
----------+---------------+---------------+------------+------------
 rq_alter | 2             | 2             | 1          | 2          
(1 row)

-- Decrease the limit
0:ALTER RESOURCE QUEUE rq_alter WITH (active_statements = 1);
ALTER QUEUE

-- Nothing changes, current holders and waiters continue as they were. However,
-- do note that now we have a situation where rsqcountvalue > rsqcountlimit.
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
 ?column? | lorrsqname | lorlocktype    | lormode       | lorgranted 
----------+------------+----------------+---------------+------------
 t        | rq_alter   | resource queue | ExclusiveLock | t          
 t        | rq_alter   | resource queue | ExclusiveLock | f          
(2 rows)
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';
 rsqname  | rsqcountlimit | rsqcountvalue | rsqwaiters | rsqholders 
----------+---------------+---------------+------------+------------
 rq_alter | 1             | 2             | 1          | 2          
(1 row)

-- The following will NOT wake up Session 2, as we would exceed the limit if we
-- were to do so.
1:CLOSE c1;
CLOSE CURSOR
0:SELECT lorusename=session_user, lorrsqname, lorlocktype, lormode, lorgranted FROM gp_toolkit.gp_locks_on_resqueue WHERE lorrsqname ='rq_alter';
 ?column? | lorrsqname | lorlocktype    | lormode       | lorgranted 
----------+------------+----------------+---------------+------------
 t        | rq_alter   | resource queue | ExclusiveLock | t          
 t        | rq_alter   | resource queue | ExclusiveLock | f          
(2 rows)
0:SELECT rsqname, rsqcountlimit, rsqcountvalue, rsqwaiters, rsqholders FROM gp_toolkit.gp_resqueue_status WHERE rsqname = 'rq_alter';
 rsqname  | rsqcountlimit | rsqcountvalue | rsqwaiters | rsqholders 
----------+---------------+---------------+------------+------------
 rq_alter | 1             | 1             | 1          | 1          
(1 row)

-- Now this will wake up Session 2
1:CLOSE c2;
CLOSE CURSOR
2<:  <... completed>
DECLARE CURSOR

1q: ... <quitting>
2q: ... <quitting>

0:DROP ROLE rq_alter_role;
DROP ROLE
0:DROP RESOURCE QUEUE rq_alter;
DROP QUEUE
